lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,409 @@
+  void runOnModule() override {
+
+    // Nothing to do if this module does not contain the "gpu.kernel_module"
+    // attribute,
+    // which is used to mark the (nested) modules created to house the GPU
+    // kernel functions
+    if (!getModule().getAttrOfType<UnitAttr>(
+            gpu::GPUDialect::getKernelModuleAttrName())) {
+      return;
+    }
+
+    // This is a module containing a GPU kernel function, we have work to do!
+
+    // Make sure the AMDGPU target is initialized.
+    LLVMInitializeAMDGPUTarget();
+    LLVMInitializeAMDGPUTargetInfo();
+    LLVMInitializeAMDGPUTargetMC();
+    LLVMInitializeAMDGPUAsmPrinter();
+
+    auto llvmModule = translateModuleToROCDLIR(getModule());
+    if (!llvmModule) {
+      return signalPassFailure();
+    }
+
+    for (auto function : getModule().getOps<FuncOp>()) {
+      if (!gpu::GPUDialect::isKernel(function)) {
+        continue;
+      }
+
+      if (failed(translateGpuKernelToHSACOAnnotation(*llvmModule, function)))
+        signalPassFailure();
+    }
+
+    // dump the MLIR
+    if (getenv("MLIR_ROCM_DUMP_AFTER_CONV_GPU_KERNEL")) {
+      std::cerr << "MLIR - After call to convert GPU Kernel" << std::endl;
+      getModule().dump();
+    }
+  }
+
+private:
+  LogicalResult translateGpuKernelToHSACOAnnotation(llvm::Module &llvmModule,
+                                                    FuncOp &function);
+
+  OwnedHSACO convertModuleToHSACO(llvm::Module &llvmModule, FuncOp &function);
+
+  OwnedHSACO emitModuleToHSACO(llvm::Module &llvmModule,
+                               llvm::TargetMachine &targetMachine);
+
+  OwnedHSACO emitModuleToHSACOForTesting(llvm::Module &llvmModule,
+                                         FuncOp &function);
+
+  rocm::HSACOGeneratorConfig config;
+};
+
+// get the "-mcpu" option string corresponding to the given AMDGPU version enum
+std::string GetMcpuOptionString(rocm::AMDGPUVersion v) {
+  switch (v) {
+  case rocm::AMDGPUVersion::GFX900:
+    return "gfx900";
+    break;
+  }
+  return "<invalid AMDGPU version>";
+}
+
+// get filename for file containing the AMDGPU version specific bitcodes
+std::string GetBitcodeFilename(rocm::AMDGPUVersion v) {
+  switch (v) {
+  case rocm::AMDGPUVersion::GFX900:
+    return "oclc_isa_version_900.amdgcn.bc";
+    break;
+  }
+  return "<invalid AMDGPU version>";
+}
+
+// get the option string corresponding to the given HSACO version enum
+std::string GetCodeObjectOptionString(rocm::HSACOVersion v) {
+  switch (v) {
+  case rocm::HSACOVersion::V3:
+    return "-code-object-v3";
+    break;
+  }
+  return "invalid HSACO version";
+}
+
+// Gets the ROCm-Device-Libs filenames for a particular AMDGPU version.
+static std::vector<std::string>
+GetROCDLPaths(const rocm::AMDGPUVersion amdgpuVersion,
+              const std::string &rocdlDir) {
+
+  // AMDGPU version-neutral bitcodes.
+  static std::vector<std::string> *rocdlFilenames =
+      new std::vector<std::string>({"hc.amdgcn.bc", "opencl.amdgcn.bc",
```

Comment:
Let's avoid the use of static std::vectors and use something like an array of constexpr StringLiteral:
https://github.com/llvm-mirror/llvm/blob/0402afb4c1b9450302a66ed04566194db86190d9/lib/Target/AMDGPU/AMDGPUAnnotateKernelFeatures.cpp#L201